pkg/querier/queryrange/limits.go

```diff
@@ -188,41 +190,123 @@ func (l limitsMiddleware) Do(ctx context.Context, r queryrangebase.Request) (que
-	matchers, err := syntax.ExtractMatchersFromQuery(r.GetQuery())
+// getBytesReadForRequest returns the number of bytes that would be read for the query in r.
+// Since the query expression may contain multiple stream matchers, this function sums up the
+// bytes that will be read for each stream.
+// E.g. for the following query:
+//
+//	count_over_time({job="foo"}[5m]) / count_over_time({job="bar"}[5m] offset 10m)
+//
+// this function will sum the bytes read for each of the following streams, taking into account
+// individual intervals and offsets
+//   - {job="foo"}
+//   - {job="bar"}
+//
+// TODO: This function shares a lot of code with dynamicShardResolver.Shards. I think we should extract the functionality to a common function?
+func (q *querySizeLimiter) getBytesReadForRequest(ctx context.Context, r queryrangebase.Request) (uint64, error) {
+	expr, err := syntax.ParseExpr(r.GetQuery())
 	if err != nil {
-		return nil, err
+		return 0, err
+	}
+
+	matcherGroups, err := syntax.MatcherGroups(expr)
+	if err != nil {
+		return 0, err
+	}
+
+	// TODO: Set concurrency dynamically as in shardResolverForConf?
+	const maxConcurrentIndexReq = 10
+	results := make([]*stats.Stats, len(matcherGroups), len(matcherGroups))
+	if err := concurrency.ForEachJob(ctx, len(matcherGroups), maxConcurrentIndexReq, func(ctx context.Context, i int) error {
+		matchers := syntax.MatchersString(matcherGroups[i].Matchers)
+		diff := matcherGroups[i].Interval + matcherGroups[i].Offset
+		adjustedFrom := model.Time(r.GetStart()).Add(-diff)
+		if matcherGroups[i].Interval == 0 {
+			// TODO: IIUC, this is needed for instant queries.
+			//       Should we use the MaxLookBackPeriod as done in shardResolverForConf?
+			adjustedFrom = adjustedFrom.Add(-100)
+		}
+
+		adjustedThrough := model.Time(r.GetEnd()).Add(-matcherGroups[i].Offset)
+
+		resp, err := q.next.Do(ctx, &logproto.IndexStatsRequest{
+			From:     adjustedFrom,
+			Through:  adjustedThrough,
+			Matchers: matchers,
+		})
+		if err != nil {
+			return err
+		}
+
+		casted, ok := resp.(*IndexStatsResponse)
+		if !ok {
+			return fmt.Errorf("expected *IndexStatsResponse while querying index, got %T", resp)
+		}
+
+		results[i] = casted.Response
+		return nil
+	}); err != nil {
+		return 0, err
 	}
 
-	// Get Stats for this query
-	var indexStatsReq definitions.Request = &logproto.IndexStatsRequest{}
-	indexStatsReq = indexStatsReq.WithStartEnd(r.GetStart(), r.GetEnd())
-	indexStatsReq = indexStatsReq.WithQuery(syntax.MatchersString(matchers))
+	var bytesRead uint64 = 0
+	for _, r := range results {
+		bytesRead += r.GetBytes()
+	}
+
+	return bytesRead, nil
+}
 
-	resp, err := q.next.Do(ctx, indexStatsReq)
+// TODO: This is pretty similar to ShardingConfigs.ValidRange, we should probably extract the functionality to a new function
+func (q *querySizeLimiter) getSchemaCfg(r queryrangebase.Request) (config.PeriodConfig, error) {
```

Comment:
I think you can reuse `ShardingConfigs` [here](https://github.com/grafana/loki/blob/main/pkg/querier/queryrange/querysharding.go#L253-L299). We can actually extend that functionality in the case there are multiple period configs in the query and they all are tsdb (and have the `IndexStats` support).

Reply:
Done with https://github.com/grafana/loki/pull/8670/commits/c9c86101d2d60a02b2437583d9fdee4afb8fa9e2. 

> We can actually extend that functionality in the case there are multiple period configs in the query and they all are tsdb

I don't know what you mean by this. If two consecutive PeriodConfig are TSDB, combine them into a new PeriodConfig with a `From` and `Interval` adjusted to cover both original periods?